prov/efa/src/rdm/rxr_ep.c

```diff
@@ -796,6 +796,67 @@ void rxr_ep_set_extra_info(struct rxr_ep *ep)
 	ep->extra_info[0] |= RXR_EXTRA_FEATURE_RUNT;
 }
 
+/**
+ * @brief set the "use_shm_for_tx" field of rxr_ep
+ * The field is set based on various factors, including
+ * environment variables, user hints, user's fi_setopt()
+ * calls.
+ * This function should be called during call to fi_enable(),
+ * after user called fi_setopt().
+ *
+ * @param[in]	ep	endpoint to set the field
+ */
+static
+void rxr_ep_set_use_shm_for_tx(struct rxr_ep *ep)
+{
+	if (!rxr_ep_domain(ep)->shm_domain) {
+		ep->use_shm_for_tx = false;
+		return;
+	}
+
+	/* App provided hints supercede environmental variables.
+	 *
+	 * Using the shm provider comes with some overheads, particularly in the
+	 * progress engine when polling an empty completion queue, so avoid
+	 * initializing the provider if the app provides a hint that it does not
+	 * require node-local communication. We can still loopback over the EFA
+	 * device in cases where the app violates the hint and continues
+	 * communicating with node-local peers.
+	 */
+	if (ep->user_info
+	    /* If the app requires explicitly remote communication */
+	    && (ep->user_info->caps & FI_REMOTE_COMM)
+	    /* but not local communication */
+	    && !(ep->user_info->caps & FI_LOCAL_COMM)) {
+		ep->use_shm_for_tx = false;
+		return;
+	}
+
+	/* TODO Update shm provider to support HMEM */
+	if (ep->user_info->caps & FI_ATOMIC && ep->user_info->caps & FI_HMEM) {
+		ep->use_shm_for_tx = false;
+		return;
+	}
+
+	/*
+	 * shm provider must make cuda calls to transfer cuda memory.
+	 * if cuda call is not allowed, we cannot use shm for transfer.
+	 *
+	 * Note that the other two hmem interfaces supported by EFA,
+	 * AWS Neuron and Habana Synapse, have no SHM provider
+	 * support anyways, so disabling SHM will not impact them.
+	 */
+	if (ep->user_info && (ep->user_info->caps & FI_HMEM)
+	    && hmem_ops[FI_HMEM_CUDA].initialized
+	    && !ep->cuda_api_permitted) {
```

Comment:
Possible to simplify? 
```
hmem_ops[FI_HMEM_CUDA].initialized && !ep->cuda_api_permitted
```

Reply:
No. `ep->user_info && (ep->user_info->caps & FI_HMEM)` is critical because this branch only apply when user explicitly requested FI_HMEM capability.